File: notion-export-file-renamer.py
import os
import re
import zipfile
import shutil

file_regexp = r'( \w{32,32}\b)'  # 파일 이름에서 사용한 정규식 패턴입니다.

def process_markdown_files():
    for root, dirs, files in os.walk(os.getcwd(), topdown=False):
        for f in files:
            f_path = os.path.join(root, f)
            if f.endswith('.md'):  # .md 파일인지 확인합니다.
                with open(f_path, 'r', encoding='utf-8') as file:
                    content = file.read()
                substituted_content = re.sub(r'\[.*?\]\((.*?)\)', lambda x: '[' + os.path.splitext(os.path.basename(x.group(1)))[0] + '](' + os.path.splitext(os.path.basename(x.group(1)))[0] + ')', content)
                # 위의 코드는 []() 패턴을 찾아서 파일명과 동일한 조건으로 변경합니다.
                with open(f_path, 'w', encoding='utf-8') as file:
                    file.write(substituted_content)

                # 파일 내의 링크 URL에 적용한 후에 파일명 변경을 수행합니다.
                substitute_f_path = re.sub(file_regexp, '', os.path.abspath(f_path))
                os.renames(f_path, substitute_f_path)

def extract_and_replace_src():
    root_dir = os.getcwd()
    raw_folder = os.path.join(root_dir, 'raw')

    # raw 폴더에서 zip 파일 검색
    zip_files = [f for f in os.listdir(raw_folder) if f.endswith('.zip')]

    if not zip_files:
        print("No zip files found in the 'raw' folder.")
        return

    # 첫 번째 zip 파일 선택
    zip_file_path = os.path.join(raw_folder, zip_files[0])

    # 압축 해제
    extract_path = os.path.join(root_dir, 'temp_extract')
    with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
        zip_ref.extractall(extract_path)

    # Notion이 내보내기 zip을 또 다른 zip으로 감싸는 새 형식 처리
    extracted_items = os.listdir(extract_path)
    if len(extracted_items) == 1 and extracted_items[0].endswith('.zip'):
        inner_zip_path = os.path.join(extract_path, extracted_items[0])
        inner_extract_path = os.path.join(root_dir, 'temp_extract_inner')
        with zipfile.ZipFile(inner_zip_path, 'r') as zip_ref:
            zip_ref.extractall(inner_extract_path)
        shutil.rmtree(extract_path)
        extract_path = inner_extract_path

    # 기존 src 폴더 삭제
    src_path = os.path.join(root_dir, 'src')
    if os.path.exists(src_path):
        shutil.rmtree(src_path)

    # 압축 해제된 전체 폴더를 src로 이동
    os.rename(extract_path, src_path)

if __name__ == "__main__":
    extract_and_replace_src()
    process_markdown_files()

File: notion-export-file-renamer.go
package main

import (
	"archive/zip"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// 파일 이름에서 사용한 정규식 패턴입니다.
var fileRegexp = regexp.MustCompile(` \w{32}\b`)

func processMarkdownFiles() {
	err := filepath.Walk(".",
		func(path string, info os.FileInfo, err error) error {
			if err != nil {
				fmt.Println("Error walking file path", err)
				return err
			}

			// .md 파일인지 확인합니다.
			if !info.IsDir() && strings.HasSuffix(info.Name(), ".md") {
				content, err := os.ReadFile(path)
				if err != nil {
					fmt.Println("Error reading file", err)
					return err
				}

				// []() 패턴을 찾아서 파일명과 동일한 조건으로 변경합니다.
				substitutedContent := regexp.MustCompile(`\[.*?\]\((.*?)\)`).ReplaceAllFunc(content,
					func(match []byte) []byte {
						matchStr := string(match)
						matches := regexp.MustCompile(`\[.*?\]\((.*?)\)`).FindStringSubmatch(matchStr)
						if len(matches) > 1 {
							fileName := filepath.Base(matches[1])
							fileNameWithoutExt := strings.TrimSuffix(fileName, filepath.Ext(fileName))
							return []byte(fmt.Sprintf("[%s](%s)", fileNameWithoutExt, fileNameWithoutExt))
						}
						return match
					},
				)

				err = os.WriteFile(path, substitutedContent, 0644)
				if err != nil {
					fmt.Println("Error writing file", err)
					return err
				}

				// 파일 내의 링크 URL에 적용한 후에 파일명 변경을 수행합니다.
				newPath := fileRegexp.ReplaceAllString(path, "")
				if newPath != path {
					newDir := filepath.Dir(newPath)
					if _, err := os.Stat(newDir); os.IsNotExist(err) {
						err := os.MkdirAll(newDir, os.ModePerm)
						if err != nil {
							fmt.Println("Error creating directory:", err)
							return err
						}
					}

					err = os.Rename(path, newPath)
					if err != nil {
						fmt.Println("Error renaming file", err)
						return err
					}

					// 왜인지 모르겠는데 해당 처리를 해도 디렉토리가 남아 있음...(iCloud로 인한 로컬 문제일지도)
					oldDir := filepath.Dir(path)
					if _, err := os.Stat(oldDir); os.IsExist(err) {
						fmt.Println("Old dir : ", oldDir)
						err := os.RemoveAll(oldDir)
						if err != nil {
							fmt.Println("Error removing old directory:", err)
							return err
						}
					}
				}
			}

			return nil
		},
	)

	if err != nil {
		fmt.Println("Error walking file path", err)
		return
	}

	fmt.Println("Successfully processed markdown files")
}

func extractAndReplaceSrc() {
	rootDir, err := os.Getwd()
	if err != nil {
		fmt.Println("Error getting current directory:", err)
		return
	}

	rawFolder := filepath.Join(rootDir, "raw")
	files, err := os.ReadDir(rawFolder)
	if err != nil {
		fmt.Println("Error reading raw folder:", err)
		return
	}

	var zipFilePath string
	for _, file := range files {
		if !file.IsDir() && strings.HasSuffix(file.Name(), ".zip") {
			zipFilePath = filepath.Join(rawFolder, file.Name())
			break
		}
	}

	if zipFilePath == "" {
		fmt.Println("No zip files found in the 'raw' folder.")
		return
	}

	extractPath := filepath.Join(rootDir, "temp_extract")
	err = unzip(zipFilePath, extractPath)
	if err != nil {
		fmt.Println("Error extracting zip file:", err)
		return
	}

	srcPath := filepath.Join(rootDir, "src")
	if _, err := os.Stat(srcPath); !os.IsNotExist(err) {
		err = os.RemoveAll(srcPath)
		if err != nil {
			fmt.Println("Error removing existing src folder:", err)
			return
		}
	}

	err = os.Rename(extractPath, srcPath)
	if err != nil {
		fmt.Println("Error moving extracted folder to src:", err)
		return
	}

	err = os.RemoveAll(extractPath)
	if err != nil {
		fmt.Println("Error removing temporary extraction folder:", err)
		return
	}

	fmt.Println("Successfully processed file", srcPath)
}

func unzip(src string, dest string) error {
	r, err := zip.OpenReader(src)
	if err != nil {
		return err
	}

	defer func(r *zip.ReadCloser) {
		err = r.Close()
		if err != nil {
			fmt.Println("Error closing zip reader:", err)
		}
	}(r)

	for _, f := range r.File {
		fPath := filepath.Join(dest, f.Name)

		if f.FileInfo().IsDir() {
			err = os.MkdirAll(fPath, os.ModePerm)
			if err != nil {
				return err
			}
			continue
		}

		if err := os.MkdirAll(filepath.Dir(fPath), os.ModePerm); err != nil {
			return err
		}

		outFile, err := os.OpenFile(fPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, f.Mode())
		if err != nil {
			return err
		}

		rc, err := f.Open()
		if err != nil {
			_ = outFile.Close()
			return err
		}

		_, err = io.Copy(outFile, rc)

		_ = outFile.Close()
		_ = rc.Close()

		if err != nil {
			return err
		}
	}

	fmt.Println("Successfully unzipped file", src)
	return nil
}

func cleanUnusedDir(dirPath string) error {
	// 디렉토리 내부에 파일이 있는지 확인
	files, err := os.ReadDir(dirPath)
	if err != nil {
		fmt.Println("Error reading directory:", err)
		return err
	}

	// 디렉토리가 비어 있으면 삭제
	if len(files) == 0 {
		err := os.Remove(dirPath) // 비어 있는 디렉토리 삭제
		if err != nil {
			fmt.Println("Error removing empty directory:", err)
			return err
		} else {
			fmt.Println("Successfully removed empty directory:", dirPath)
		}
	} else {
		fmt.Println("Directory is not empty:", dirPath)
	}

	return nil
}

func main() {
	extractAndReplaceSrc()
	processMarkdownFiles()
}
